Linter output no longer shows up under WARN or INFO log levels

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Linter output is not displayed under log level WARN or INFO (it was displayed under WARN in previous versions). It is now only displayed under the DEBUG log level.

### Expected Behavior

In previous versions super-linter would display the output from the linters in WARN log level so that we could see what the linters are complaining about and fix it. Now this information is only displayed under DEBUG logging level which produces tens of thousands or even hundreds of thousands of lines. It's impractical to have to sift through all of that just to figure out what's wrong every time we want to lint something.

### Super-Linter version

```markdown
I can't figure out how to get the version from inside the container and it doesn't seem to output
it when running. It's either 6.0.0 or 5.7.2. I pulled it down yesterday (02/01/2024) as the latest
docker image, so whatever version that would be.
```


### Relevant log output

```shell
Note that linters fail but don't show the output from the linters indicating what they are complaining about.
We use this linter automatically in places but we also use it as a manual tool for reviewing one off scripts
before they are submitted, and with the linter output missing from WARN level logging this is a major
issue for us. Scraping through a mountain of DEBUG log output to find the linter output is a waste of time.

Also, all of this linter version info being output just below the ascii art never used to be there when using
LOG_LEVEL=WARN. That all seems like it should only be output when LOG_LEVEL=INFO is in use.
LOG_LEVEL=WARN should only show linting failures and output from the linters like it used to
(as well as any fatal error conditions).

Logging levels in general seem to have gotten pretty messed up with this last update.

--------------------------------------------------------------------------------

                              /@@#///////@@/(@//@%/(@.@(       @@
                          @@//////////////////////////////#*  @@@
                        @////@//(///////////@@@@@///@//@/@**//@@(
                      @///////@///////////////@@@@    (           @,
                     @/(&/@////////////////////                     @
                    @////////////////////////@@                      @
                  @%////////(//////////%/////&@            @@       *,@           ______________
             @@@@@/@/#/////(&//////////////////                       .@         /              \
        *@@@@@.    .%///(//@//////////////////&.   .@@,                 @%      / Don't mind me  \
      @@%           .&@&&/@.@//&/////(//////////    @@@@@@@@@         .. &@    / I'm just looking \
    @@%               @@@@@   @&/////////////////#   @/       V  @@/ ,@@@ @   <  for some trash... |
@@@%                   @@@@        .%@@@@//////#@ @   @@         @     .,.     \__________________/
                                          @@@/@(  (@@@@% @/\      %
                                           @@@@(    .     .@@/\   #
                                             @                  %@%

--------------------------------------------------------------------------------
chktex: ChkTeX v1.7.8 - Copyright 1995-96 Jens T. Berger Thielemann.
Compiled with POSIX extended regex support.
rubocop: 1.59.0
perl:
This is perl 5, version 38, subversion 2 (v5.38.2) built for x86_64-linux-thread-multi

Copyright 1987-2023, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at https://www.perl.org/, the Perl Home Page.
xmllint: xmllint: using libxml version 21106
   compiled with: Threads Tree Output Push Reader Patterns Writer SAXv1 HTTP DTDValid HTML C14N Catalog XPath XPointer XInclude Iconv ISO8859X Unicode Regexps Automata Schemas Schematron Modules Debug Zlib Lzma
ansible-lint: ansible-lint 6.22.2 using ansible-core:2.16.3 ansible-compat:4.1.11 ruamel-yaml:0.18.5 ruamel-yaml-clib:0.2.8
eslint: v8.56.0
markdownlint: 0.39.0
snakefmt: snakefmt, version 0.9.0
cpplint: Cpplint fork (https://github.com/cpplint/cpplint)
cpplint 1.6.1
Python 3.12.1 (main, Jan 27 2024, 06:36:14) [GCC 13.2.1 20231014]
editorconfig-checker: 2.7.2
gitleaks: v8.18.1
phpstan: PHPStan - PHP Static Analysis Tool 1.10.48
dotenv-linter: dotenv-linter 3.3.0
asl-validator: 3.8.2
actionlint: 1.6.26
installed by building from source
built with go1.21.1 compiler for linux/amd64
tekton-lint: 1.0.0-beta.9
terrascan: version: v1.18.11
terraform: Terraform v1.7.1
on linux_amd64
2024/01/30 20:02:15 traces export: Post "http:///dev/otel-grpc.sock": http: no Host in request URL
2024/01/30 20:02:15 traces export: Post "http:///dev/otel-grpc.sock": http: no Host in request URL
clippy: clippy 0.1.72
R: R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-musl (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under the terms of the
GNU General Public License versions 2 or 3.
For more information about these matters see
https://www.gnu.org/licenses/.
rustfmt: rustfmt 1.6.0-stable ( )
shellcheck: ShellCheck - shell script analysis tool
version: 0.9.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
jscpd: 3.5.10
black: black, 24.1.1 (compiled: no)
Python (CPython) 3.12.1
php: PHP 8.2.15 (cli) (built: Jan 18 2024 16:40:05) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.15, Copyright (c) Zend Technologies
textlint: v13.4.1
npm-groovy-lint: GroovyLint: Successfully processed CodeNarc:
CodeNarc version 3.3.0
npm-groovy-lint version 14.2.0

Embeds:
CodeNarc version 3.3.0
- Groovy version 3.0.9 (superlite)
renovate-config-validator: (node:399) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
37.161.0
golangci-lint: golangci-lint has version v1.55.2 built with go1.21.3 from e3c2265f on 2023-11-03T13:27:54Z
kubeconform: v0.6.4
golangci-lint: golangci-lint has version v1.55.2 built with go1.21.3 from e3c2265f on 2023-11-03T13:27:54Z
gherkin-lint: Version command not supported
eslint: v8.56.0
scalafmt: scalafmt 3.7.17
stylelint: 15.11.0
sqlfluff: sqlfluff, version 2.3.5
bash-exec: Version command not supported
google-java-format: google-java-format: Version 1.19.2
phpcs: PHP_CodeSniffer version 3.8.0 (stable) by Squiz and PHPCSStandards
terragrunt: terragrunt version v0.54.22
psalm: Psalm 5.20.0@3f284e96c9d9be6fe6b15c79416e1d1903dcfef4
pylint: pylint 3.0.3
astroid 3.0.2
Python 3.12.1 (main, Jan 27 2024, 06:36:14) [GCC 13.2.1 20231014]
shfmt: v3.7.0
snakemake: 8.4.1
arm-ttk: ModuleVersion = 0.24
pwsh: PowerShell 7.4.1
mypy: mypy 1.8.0 (compiled: yes)
eslint: v8.56.0
coffeelint: 5.2.11
tflint: TFLint version 0.50.2
+ ruleset.terraform (0.5.0-bundled)
cfn-lint: cfn-lint 0.85.0
flake8: 7.0.0 (mccabe: 0.7.0, pycodestyle: 2.11.1, pyflakes: 3.2.0) CPython 3.12.1 on
Linux
spectral: 6.11.0
clj-kondo: clj-kondo v2023.12.15
eslint: v8.56.0
ts-standard: 12.0.2
eslint: v8.56.0
protolint: protolint version 0.47.5(80ba08f)
raku: Welcome to Rakudo™ v2023.11.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2023.11.
hadolint: Haskell Dockerfile Linter 2.12.0
checkstyle: Checkstyle version: 10.13.0
clang-format: clang-format version 17.0.5 (https://github.com/llvm/llvm-project.git 98bfdac5ce82d1679f8af9a57501471812ab68d7)
ktlint: 1.1.1
yamllint: yamllint 1.33.0
dart: Dart SDK version: 3.2.5 (stable) (Tue Jan 16 15:02:13 2024 +0000) on "linux_x64"
isort:
                 _                 _
                (_) ___  ___  _ __| |_
                | |/ _/ / _ \/ '__  _/
                | |\__ \/\_\/| |  | |_
                |_|\___/\___/\_/   \_/

      isort your imports, so you don't have to.

                    VERSION 5.13.2
dotnet: 6.0.418
eslint: v8.56.0
rustfmt: rustfmt 1.6.0-stable ( )
htmlhint: 1.1.4
rustfmt: rustfmt 1.6.0-stable ( )
sql-lint: 1.0.0
checkov: 3.2.1
lua: Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio
standard: 17.1.0
2024-02-02 20:33:08 [ERROR]   Errors found in PYTHON_BLACK
2024-02-02 20:33:08 [ERROR]   Errors found in PYTHON_FLAKE8
2024-02-02 20:33:08 [ERROR]   Super-linter detected linting errors
```


### Steps To Reproduce

Run an older version of super-linter and lint some code which fails linting with LOG_LEVEL=WARN enabled, notice that it shows you the output from the linter indicating what it's not liking. Then run the latest version of super-linter in the same way and notice that it'll tell you that linters failed but it won't show you what they are complaining about like it did in previous versions. 

### Anything else?

With LOG_LEVEL=WARN it used to only output the ascii art at the top and any linting specific errors. Now it floods the screen with a bunch of linter version info which seems informational and should probably only be output when LOG_LEVEL=INFO is in use. It seems like log levels got messed up in general in this latest update.

Just to summarize this...

LOG_LEVEL=WARN displays errors like this but no linter output (we used to see linter output when running with WARN but we don't anymore).

2024-02-02 21:36:17 [ERROR]   Errors found in PYTHON_BLACK
2024-02-02 21:36:17 [ERROR]   Errors found in PYTHON_FLAKE8
2024-02-02 21:36:17 [ERROR]   Super-linter detected linting errors

LOG_LEVEL=INFO doesn't display the above ERROR lines at all, even though I'm linting the same file with it and it should be displaying those.

Not setting LOG_LEVEL at all shows me all the stuff I want to see so we'll just use that as a workaround for now on our end.

I would expected a LOG_LEVEL=INFO to display anything INFO or above but those ERROR lines above get hidden when I run it that way. There seems to be some sort of weird bug here that's probably worth investigating.

--- a/Makefile
+++ b/Makefile
@@ -4,7 +4,7 @@
 all: info docker test ## Run all targets.
 
 .PHONY: test
-test: info validate-container-image-labels test-lib inspec lint-codebase test-default-config-files test-find lint-subset-files test-custom-ssl-cert test-non-default-workdir test-git-flags test-non-default-home-directory test-log-level test-linters ## Run the test suite
+test: info validate-container-image-labels test-lib inspec lint-codebase test-default-config-files test-find lint-subset-files test-custom-ssl-cert test-non-default-workdir test-git-flags test-non-default-home-directory test-log-level test-linters-expect-failure-log-level-notice test-linters ## Run the test suite
 
 # if this session isn't interactive, then we don't want to allocate a
 # TTY, which would fail, but if it is interactive, we do want to attach
@@ -315,6 +315,12 @@ test-log-level: ## Run a test to check if there are conflicts with the LOG_LEVEL
 	$(CURDIR)/test/run-super-linter-tests.sh \
 		$(SUPER_LINTER_TEST_CONTAINER_URL) \
 		"run_test_cases_log_level"
+
+.phony: test-linters-expect-failure-log-level-notice
+test-linters-expect-failure-log-level-notice: ## Run the linters test suite expecting failures with a LOG_LEVEL set to NOTICE
+	$(CURDIR)/test/run-super-linter-tests.sh \
+		$(SUPER_LINTER_TEST_CONTAINER_URL) \
+		"run_test_cases_expect_failure_notice_log"
 
 .phony: build-dev-container-image
 build-dev-container-image: ## Build commit linter container image

--- a/lib/functions/detectFiles.sh
+++ b/lib/functions/detectFiles.sh
@@ -318,7 +318,7 @@ function RunAdditionalInstalls() {
     local MYPY_CACHE_DIRECTORY_PATH
     MYPY_CACHE_DIRECTORY_PATH="${GITHUB_WORKSPACE}/.mypy_cache"
     debug "Create MyPy cache directory: ${MYPY_CACHE_DIRECTORY_PATH}"
-    mkdir -v "${MYPY_CACHE_DIRECTORY_PATH}"
+    mkdir -p "${MYPY_CACHE_DIRECTORY_PATH}"
   fi
 
   ###############################

--- a/lib/functions/worker.sh
+++ b/lib/functions/worker.sh
@@ -193,7 +193,16 @@ function LintCodebase() {
   fi
 
   if [ -n "${STDOUT_LINTER}" ]; then
-    info "Command output for ${FILE_TYPE}:\n------\n${STDOUT_LINTER}\n------"
+    local STDOUT_LINTER_LOG_MESSAGE
+    STDOUT_LINTER_LOG_MESSAGE="Command output for ${FILE_TYPE}:\n------\n${STDOUT_LINTER}\n------"
+    info "${STDOUT_LINTER_LOG_MESSAGE}"
+
+    if [ ${PARALLEL_COMMAND_RETURN_CODE} -ne 0 ]; then
+      local STDOUT_LINTER_FILE_PATH
+      STDOUT_LINTER_FILE_PATH="/tmp/super-linter-parallel-stdout-${FILE_TYPE}"
+      debug "Saving stdout for ${FILE_TYPE} to ${STDOUT_LINTER_FILE_PATH} in case we need it later"
+      printf '%s\n' "${STDOUT_LINTER_LOG_MESSAGE}" >"${STDOUT_LINTER_FILE_PATH}"
+    fi
   else
     debug "Stdout for ${FILE_TYPE} is empty"
   fi
@@ -204,7 +213,15 @@ function LintCodebase() {
   fi
 
   if [ -n "${STDERR_LINTER}" ]; then
-    info "Stderr contents for ${FILE_TYPE}:\n------\n${STDERR_LINTER}\n------"
+    local STDERR_LINTER_LOG_MESSAGE
+    STDERR_LINTER_LOG_MESSAGE="Stderr contents for ${FILE_TYPE}:\n------\n${STDERR_LINTER}\n------"
+    info "${STDERR_LINTER_LOG_MESSAGE}"
+    if [ ${PARALLEL_COMMAND_RETURN_CODE} -ne 0 ]; then
+      local STDERR_LINTER_FILE_PATH
+      STDERR_LINTER_FILE_PATH="/tmp/super-linter-parallel-stderr-${FILE_TYPE}"
+      debug "Saving stderr for ${FILE_TYPE} to ${STDERR_LINTER_FILE_PATH} in case we need it later"
+      printf '%s\n' "${STDERR_LINTER_LOG_MESSAGE}" >"${STDERR_LINTER_FILE_PATH}"
+    fi
   else
     debug "Stderr for ${FILE_TYPE} is empty"
   fi

--- a/lib/linter.sh
+++ b/lib/linter.sh
@@ -626,6 +626,25 @@ Footer() {
 
       if [[ ${ERROR_COUNTER} -ne 0 ]]; then
         error "Errors found in ${LANGUAGE}"
+        # Print output as error in case users disabled the INFO level so they
+        # get feedback
+        if [[ "${LOG_VERBOSE}" != "true" ]]; then
+          local STDOUT_LINTER_FILE_PATH
+          STDOUT_LINTER_FILE_PATH="/tmp/super-linter-parallel-stdout-${LANGUAGE}"
+          if [[ -e "${STDOUT_LINTER_FILE_PATH}" ]]; then
+            error "$(cat "${STDOUT_LINTER_FILE_PATH}")"
+          else
+            debug "Stdout output file path for ${LANGUAGE} (${STDOUT_LINTER_FILE_PATH}) doesn't exist"
+          fi
+
+          local STDERR_LINTER_FILE_PATH
+          STDERR_LINTER_FILE_PATH="/tmp/super-linter-parallel-stderr-${LANGUAGE}"
+          if [[ -e "${STDERR_LINTER_FILE_PATH}" ]]; then
+            error "$(cat "${STDERR_LINTER_FILE_PATH}")"
+          else
+            debug "Stderr output file path for ${LANGUAGE} (${STDERR_LINTER_FILE_PATH}) doesn't exist"
+          fi
+        fi
         CallStatusAPI "${LANGUAGE}" "error"
         SUPER_LINTER_EXIT_CODE=1
         debug "Setting super-linter exit code to ${SUPER_LINTER_EXIT_CODE} because there were errors for ${LANGUAGE}"
@@ -687,13 +706,13 @@ cleanup() {
   local -ri EXIT_CODE=$?
 
   debug "Removing temporary files and directories"
-  rm -rfv \
+  rm -rf \
     "${GITHUB_WORKSPACE}/.mypy_cache" \
     "${GITHUB_WORKSPACE}/logback.log"
 
   if [ "${SUPER_LINTER_COPIED_R_LINTER_RULES_FILE}" == "true" ]; then
     debug "Deleting ${R_RULES_FILE_PATH_IN_ROOT} because super-linter created it."
-    rm -rfv "${R_RULES_FILE_PATH_IN_ROOT}"
+    rm -rf "${R_RULES_FILE_PATH_IN_ROOT}"
   fi
 
   # Define this variable here so we can rely on it as soon as possible
@@ -703,7 +722,6 @@ cleanup() {
     debug "Moving log file from ${LOG_TEMP} to ${LOG_FILE_PATH}"
     mv \
       --force \
-      --verbose \
       "${LOG_TEMP}" "${LOG_FILE_PATH}"
   else
     debug "Skipping the moving of the log file from ${LOG_TEMP} to ${LOG_FILE_PATH}"

--- a/test/run-super-linter-tests.sh
+++ b/test/run-super-linter-tests.sh
@@ -23,6 +23,11 @@ run_test_cases_log_level() {
   LOG_LEVEL="NOTICE"
 }
 
+run_test_cases_expect_failure_notice_log() {
+  run_test_cases_expect_failure
+  LOG_LEVEL="NOTICE"
+}
+
 run_test_cases_non_default_home() {
   run_test_cases_expect_success
   COMMAND_TO_RUN+=(-e HOME=/tmp)

